[iOS] Print preview should use images instead of PDFs

rr-codes opened the pull request:
#### d56a44cbd116ed60bde26f13beaf33969327feb6
<pre>
[iOS] Print preview should use images instead of PDFs
<a href="https://bugs.webkit.org/show_bug.cgi?id=220815">https://bugs.webkit.org/show_bug.cgi?id=220815</a>
rdar://65129050

Reviewed by Aditya Keerthi.

When generating the print preview, the system will now use an image-backed preview
instead of a PDF-backed one for better performance, if a responsive print rendering
quality was requested. The logic added is analogous with how the PDF generation works.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/_WKWebViewPrintFormatter.h:
* Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
(-[_WKWebViewPrintFormatter _shouldUseResponsivePrintRendering]):
(-[_WKWebViewPrintFormatter _printPreviewImage]):
(-[_WKWebViewPrintFormatter _setPrintPreviewImage:]):
(-[_WKWebViewPrintFormatter _waitForPrintedDocumentOrImage]):
(-[_WKWebViewPrintFormatter _invalidatePrintRenderingState]):
(-[_WKWebViewPrintFormatter _recalcPageCount]):
(-[_WKWebViewPrintFormatter drawInRect:forPageAtIndex:]):
(-[_WKWebViewPrintFormatter _drawInRectUsingBitmap:forPageAtIndex:]):
(-[_WKWebViewPrintFormatter _drawInRectUsingPDF:forPageAtIndex:]):
(-[_WKWebViewPrintFormatter _waitForPrintedDocument]): Deleted.
* Source/WebKit/UIProcess/_WKWebViewPrintFormatterInternal.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[_WKPrintFormattingAttributes initWithPageCount:frameID:printInfo:]):
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView _resetPrintRenderingState]):
(-[WKContentView _resetPrintingState]):
(-[WKContentView _attributesForPrintFormatter:]):
(-[WKContentView _wk_pageCountForPrintFormatter:]):
(-[WKContentView _createImage:isPrintingOnBackgroundThread:printFormatter:]):
(-[WKContentView _createPDF:isPrintingOnBackgroundThread:printFormatter:]):
(-[WKContentView _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:]):
(-[WKContentView _wk_requestDocumentForPrintFormatter:]):
(-[WKContentView _waitForDrawToImageCallbackForPrintFormatterIfNeeded:]):
(-[WKContentView _wk_requestImageForPrintFormatter:]):
(-[WKContentView _wk_supportsRenderingBitmap]):
* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView _wk_supportsRenderingBitmap]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::drawToImage):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawPagesToPDFImpl):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::drawToImage):

Canonical link: <a href="https://commits.webkit.org/260644@main">https://commits.webkit.org/260644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c3129aed193d3b610e3346d0e4daca56d82812b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17793 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117801 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118003 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19244 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9070 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100941 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114458 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14437 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97672 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42423 "Found 1 new test failure: webgl/2.0.y/conformance2/extensions/promoted-extensions-in-shaders.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29311 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84269 "Found 2 new API test failures: /TestWebKit:WebKit.OnDeviceChangeCrash, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10597 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30662 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11356 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7573 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50252 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7357 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12945 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatter.h:
```diff
@@ -35,7 +35,6 @@ WK_CLASS_AVAILABLE(ios(8.0))
 
 @property (nonatomic, strong) _WKFrameHandle *frameToPrint;
 @property (nonatomic) BOOL snapshotFirstPage WK_API_AVAILABLE(ios(11.0));
-
```
this seems wrong

hortont424 commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
```diff
@@ -71,6 +73,20 @@ - (WKWebView *)_webView
     return static_cast<WKWebView *>(view);
 }
 
+- (BOOL)_shouldUseResponsivePrintRendering
```
Let's keep the confusing "responsive" name as confined as possible (not here, for example)

hortont424 commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatterInternal.h:
```diff
@@ -33,16 +33,24 @@
 @protocol _WKWebViewPrintProvider <NSObject>
+@property (nonatomic, readonly) BOOL _wk_printFormatterRequiresMainThread;
+
+- (BOOL)_wk_supportsRenderingBitmap;
```
Why this extra method instead of just checking if the provider implements the optional method?

hortont424 commented on Source/WebKit/UIProcess/_WKWebViewPrintFormatterInternal.h:
```diff
@@ -33,16 +33,24 @@
 @end
 
 @protocol _WKWebViewPrintProvider <NSObject>
+@property (nonatomic, readonly) BOOL _wk_printFormatterRequiresMainThread;
```
Newline above here, I think

hortont424 commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4278,6 +4279,63 @@ static bool selectionIsInsideFixedPositionContainer(Frame& frame)
     reply(pageRects.size());
 }
 
+void WebPage::drawToImage(WebCore::FrameIdentifier frameID, const PrintInfo& printInfo, size_t pageCount, CompletionHandler<void(WebKit::ShareableBitmapHandle&&)>&& reply)
```
In reality this is *print*-to-image, right? not draw-to-image (like a software snapshot). It does printing layout and print stylesheets and whatnot?

rr-codes replied:
Ya; I named it this for consistency with `WebPage::drawToPDFiOS` which does very similar things

pxlcoder replied:
Correct. I believe `drawToPDF` also suffers from this naming problem – so we could rename both (or do that one in a follow-up, if it's going to greatly increase the size of the change).

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -965,22 +1001,68 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
     if (!pageCount)
+        return nil;
+
+    _WKPrintFormattingAttributes *attributes = [[_WKPrintFormattingAttributes alloc] initWithPageCount:pageCount frameID:frameID printInfo:printInfo];
```
This is a memory leak.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -156,7 +183,8 @@ @implementation WKContentView {
     RetainPtr<NSMutableSet> _pendingBackgroundPrintFormatters;
-    IPC::Connection::AsyncReplyID _pdfPrintCallbackID;
+    IPC::Connection::AsyncReplyID _printRenderingCallbackID;
+    _WKPrintRenderingState _printRenderingState;
```
Since this variable simply describes the nature of the callback ID above, I don't think we need `_WKPrintRenderingStateNone`. A zero `_printRenderingCallbackID` would imply that.

I would also rename `_printRenderingState` to reflect that it describes the callback, since it doesn't appear as though it's consistently updated everywhere (particularly in the background thread cases).

rr-codes replied:
@pxlcoder any suggestions as to what to rename `_printRenderingState` to? `_printRenderingCallbackState`?

pxlcoder replied:
`_printRenderingCallbackType`?

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -965,22 +1001,68 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+    else
+        [self _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:printFormatter];
+
+    RetainPtr attributes = [self _attributesForPrintFormatter:printFormatter];
```
Nit: `auto`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -1016,14 +1105,75 @@ - (void)_waitForDrawToPDFCallbackForPrintFormatterIfNeeded:(_WKWebViewPrintForma
 {
+    bool isPrintingOnBackgroundThread = !isMainRunLoop();
+
+    RetainPtr attributes = [self _attributesForPrintFormatter:printFormatter];
```
Nit: `auto`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -965,22 +1001,68 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
 
-        if (!_pendingBackgroundPrintFormatters)
-            _pendingBackgroundPrintFormatters = adoptNS([[NSMutableSet alloc] init]);
+- (void)_createImage:(_WKPrintFormattingAttributes *)formatterAttributes isPrintingOnBackgroundThread:(BOOL)isPrintingOnBackgroundThread printFormatter:(_WKWebViewPrintFormatter *)printFormatter
```
Why does `isPrintingOnBackgroundThread` need to be a parameter? Same question for the method below.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -965,22 +992,71 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+    ensureOnMainRunLoop([formatterAttributes = retainPtr(formatterAttributes), isPrintingOnBackgroundThread, printFormatter = retainPtr(printFormatter), retainedSelf = retainPtr(self)] {
         // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
-        auto callbackID = retainedSelf->_page->drawToPDFiOS(frameID, printInfo, pageCount, [isPrintingOnBackgroundThread, printFormatter, retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) mutable {
+        auto callbackID = retainedSelf->_page->drawToPDFiOS([formatterAttributes frameID], [formatterAttributes printInfo], [formatterAttributes pageCount], [isPrintingOnBackgroundThread, printFormatter, formatterAttributes, retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) mutable {
```
I don't think we need to capture `formatterAttributes`?

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -900,19 +927,17 @@ - (BOOL)_wk_printFormatterRequiresMainThread
 {
     bool isPrintingOnBackgroundThread = !isMainRunLoop();
 
-    [self _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:printFormatter];
```
Is it safe to remove this? I see call sites of this method where we no longer call `_waitForDraw...`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -965,22 +992,71 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
+    if (printFormatter._shouldDrawUsingBitmap)
+        [self _waitForDrawToImageCallbackForPrintFormatterIfNeeded:printFormatter];
+    else
+        [self _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:printFormatter];
```
Do we need an `if` here? The methods are "if needed", so they should do nothing if they don't apply?

rr-codes committed:
[iOS] Print preview should use images instead of PDFs
https://bugs.webkit.org/show_bug.cgi?id=220815
rdar://65129050

Reviewed by Aditya Keerthi.

When generating the print preview, the system will now use an image-backed preview
instead of a PDF-backed one for better performance, if a responsive print rendering
quality was requested. The logic added is analogous with how the PDF generation works.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/_WKWebViewPrintFormatter.h:
* Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm:
(-[_WKWebViewPrintFormatter _shouldUseResponsivePrintRendering]):
(-[_WKWebViewPrintFormatter _printPreviewImage]):
(-[_WKWebViewPrintFormatter _setPrintPreviewImage:]):
(-[_WKWebViewPrintFormatter _waitForPrintedDocumentOrImage]):
(-[_WKWebViewPrintFormatter _invalidatePrintRenderingState]):
(-[_WKWebViewPrintFormatter _recalcPageCount]):
(-[_WKWebViewPrintFormatter drawInRect:forPageAtIndex:]):
(-[_WKWebViewPrintFormatter _drawInRectUsingBitmap:forPageAtIndex:]):
(-[_WKWebViewPrintFormatter _drawInRectUsingPDF:forPageAtIndex:]):
(-[_WKWebViewPrintFormatter _waitForPrintedDocument]): Deleted.
* Source/WebKit/UIProcess/_WKWebViewPrintFormatterInternal.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[_WKPrintFormattingAttributes initWithPageCount:frameID:printInfo:]):
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView _resetPrintRenderingState]):
(-[WKContentView _resetPrintingState]):
(-[WKContentView _attributesForPrintFormatter:]):
(-[WKContentView _wk_pageCountForPrintFormatter:]):
(-[WKContentView _createImage:isPrintingOnBackgroundThread:printFormatter:]):
(-[WKContentView _createPDF:isPrintingOnBackgroundThread:printFormatter:]):
(-[WKContentView _waitForDrawToPDFCallbackForPrintFormatterIfNeeded:]):
(-[WKContentView _wk_requestDocumentForPrintFormatter:]):
(-[WKContentView _waitForDrawToImageCallbackForPrintFormatterIfNeeded:]):
(-[WKContentView _wk_requestImageForPrintFormatter:]):
(-[WKContentView _wk_supportsRenderingBitmap]):
* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView _wk_supportsRenderingBitmap]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::drawToImage):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawPagesToPDFImpl):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::drawToImage):

Canonical link: https://commits.webkit.org/260644@main
